nupic/research/frameworks/pytorch/imagenet/mixins/maxup.py

```diff
@@ -0,0 +1,142 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2020, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+
+
+class MaxupStandard(object):
+    """
+    Maxup: applies m transformations to each sample;
+    concatenates into m different batches;
+    learns from batch with worst loss.
+
+    Paper: https://arxiv.org/pdf/2002.09024.pdf
+
+    Requires ComplexLoss Mixin
+    """
+    def __init__(self):
+        super().__init__()
+
+        self.execution_order["calculate_batch_loss"] = [
+            "MaxupStandard.calculate_batch_loss"
+        ]
+
+    def calculate_batch_loss(self, data, target, async_gpu=True):
+        """
+        :param data: input to the training function, as specified by dataloader
+        :param target: target to be matched by model, as specified by dataloader
+        :param async_gpu: define whether or not to use
+                          asynchronous GPU copies when the memory is pinned
+        """
+
+        if len(data.shape) < 5:
+            raise ValueError("Define replicas_per_sample > 1")
+
+        target = target.to(self.device, non_blocking=async_gpu)
```

Comment:
It looks like it would be simpler to move this optimization to the callers in `ComplexLoss. _train_model`
https://github.com/numenta/nupic.research/blob/master/nupic/research/frameworks/pytorch/imagenet/mixins/complex_loss.py#L71

Reply:
The ComplexLoss is reused by other mixins, including Knowledge Distillation. In the Knowledge Distillation case, I don't need the target in some cases -  hence I get extra memory if I don't move target to the GPU, which is the reason I left this step to be done inside the calculate_batch_loss function.

Reply:
Makes sense.

Reply:
In that case I think you should use the `async_gpu` local variable instead of forcing it to be true when calling `calculate_batch_loss`. See https://github.com/numenta/nupic.research/blob/master/nupic/research/frameworks/pytorch/imagenet/mixins/complex_loss.py#L79